src/sources/spotify.rs

```diff
@@ -0,0 +1,158 @@
+        spotify.request_token().await?;
+
+        Ok(spotify)
+    }
+
+    pub async fn extract(
+        spotify: &ClientCredsSpotify,
+        query: &str,
+    ) -> Result<QueryType, ParrotError> {
+        let re = Regex::new(r"spotify.com/(?P<media_type>.+)/(?P<media_id>.*?)(?:\?|$)").unwrap();
+        let captures = re
+            .captures(query)
+            .ok_or(ParrotError::Other(SPOTIFY_INVALID_QUERY))?;
+
+        let media_type = captures
+            .name("media_type")
+            .ok_or(ParrotError::Other(SPOTIFY_INVALID_QUERY))?
+            .as_str();
+
+        let media_type = MediaType::from_str(media_type)
+            .map_err(|_| ParrotError::Other(SPOTIFY_INVALID_QUERY))?;
+
+        let media_id = captures
+            .name("media_id")
+            .ok_or(ParrotError::Other(SPOTIFY_INVALID_QUERY))?
+            .as_str();
+
+        match media_type {
+            MediaType::Track => Self::get_track_info(spotify, media_id).await,
+            MediaType::Album => Self::get_album_info(spotify, media_id).await,
+            MediaType::Playlist => Self::get_playlist_info(spotify, media_id).await,
+        }
+    }
+
+    async fn get_track_info(
+        spotify: &ClientCredsSpotify,
+        id: &str,
+    ) -> Result<QueryType, ParrotError> {
+        let track_id = TrackId::from_id(id)
+            .map_err(|_| ParrotError::Other("track ID contains invalid characters"))?;
+
+        let track = spotify
+            .track(&track_id)
+            .await
+            .map_err(|_| ParrotError::Other("failed to fetch track"))?;
+
+        let artist_names = Self::join_artist_names(&track.artists);
+
+        let query = Self::build_query(&artist_names, &track.name);
+        Ok(QueryType::Keywords(query))
+    }
+
+    async fn get_album_info(
+        spotify: &ClientCredsSpotify,
+        id: &str,
+    ) -> Result<QueryType, ParrotError> {
+        let album_id = AlbumId::from_id(id)
+            .map_err(|_| ParrotError::Other("album ID contains invalid characters"))?;
+
+        let album = spotify
+            .album(&album_id)
+            .await
+            .map_err(|_| ParrotError::Other("failed to fetch album"))?;
+
+        let artist_names = Self::join_artist_names(&album.artists);
+
+        let query_list: Vec<String> = album
+            .tracks
+            .items
+            .iter()
+            .map(|track| Self::build_query(&artist_names, &track.name))
+            .collect();
+
+        Ok(QueryType::KeywordList(query_list))
+    }
+
+    async fn get_playlist_info(
+        spotify: &ClientCredsSpotify,
+        id: &str,
+    ) -> Result<QueryType, ParrotError> {
+        let playlist_id = PlaylistId::from_id(id)
+            .map_err(|_| ParrotError::Other("playlist ID contains invalid characters"))?;
+
+        let playlist = spotify
+            .playlist(
+                &playlist_id,
+                Some(""),
+                Some(&Market::Country(Country::UnitedStates)),
```

Comment:
This playlist method requires `fields` and `market` parameters.

Reply:
But why are we defaulting to UnitedStates?

Reply:
Aren't these options @aquelemiguel ? Can't you just pass `None` to both instead of `Some("")` and the US market one.

Reply:
In fact:
https://docs.rs/rspotify/0.11.4/rspotify/clients/base/trait.BaseClient.html#method.playlist

![image](https://user-images.githubusercontent.com/16060539/158020553-2908f5b8-127b-423f-ba33-ddd6c6c0d2ca.png)

Maybe you can literally pass "from_token" and it will use the market associated with the token (for US users will be US)

Reply:
Good thinking but I'm afraid with our current OAuth flow we can't fetch user-related info.
I've tried your suggestion and it failed to fetch playlists.

![image](https://user-images.githubusercontent.com/5618493/158023453-d60cac75-278d-44c5-977b-f2998631dbec.png)
See more: https://developer.spotify.com/documentation/general/guides/authorization/

Reply:
Alright, let's go with US then. What are the implications? I did not know Spotify was split into markets? Does it mean some musics from Japan won't be available for the US market for example, or vice-versa?

Reply:
Every media content is split into markets. Which is why I raised this concern. Maybe we can use a default, but allowing users to change this. It can be in another issue though.

Reply:
I'm passing both as None for now. If we ever change the OAuth flow, we can use this Market::FromToken thing.